controllers/mysqlcluster_controller.go

```diff
@@ -498,254 +514,269 @@ func (r *MySQLClusterReconciler) reconcileV1Service(ctx context.Context, req ctr
+				WithSecretName(cluster.MyCnfSecretName()).
+				WithDefaultMode(0644)),
+		corev1ac.Volume().
+			WithName(constants.GRPCSecretVolumeName).
+			WithSecret(corev1ac.SecretVolumeSource().
+				WithSecretName(cluster.GRPCSecretName()).
+				WithDefaultMode(0644)),
+	)
 
-	var orig, updated *appsv1.StatefulSetSpec
-	result, err := ctrl.CreateOrUpdate(ctx, r.Client, sts, func() error {
-		if debugController {
-			orig = sts.Spec.DeepCopy()
-		}
+	if !cluster.Spec.DisableSlowQueryLogContainer {
+		podSpec.WithVolumes(
+			corev1ac.Volume().
+				WithName(constants.SlowQueryLogAgentConfigVolumeName).
+				WithConfigMap(corev1ac.ConfigMapVolumeSource().
+					WithName(cluster.SlowQueryLogAgentConfigMapName()).
+					WithDefaultMode(0644)),
+		)
+	}
 
-		sts.Labels = mergeMap(sts.Labels, labelSet(cluster, false))
+	containers := make([]*corev1ac.ContainerApplyConfiguration, 0, 4)
 
-		sts.Spec.Replicas = pointer.Int32(cluster.Spec.Replicas)
-		sts.Spec.Selector = &metav1.LabelSelector{
-			MatchLabels: labelSet(cluster, false),
-		}
-		sts.Spec.PodManagementPolicy = appsv1.ParallelPodManagement
-		sts.Spec.UpdateStrategy = appsv1.StatefulSetUpdateStrategy{
-			Type: appsv1.RollingUpdateStatefulSetStrategyType,
-		}
-		sts.Spec.ServiceName = cluster.HeadlessServiceName()
-
-		sts.Spec.VolumeClaimTemplates = make([]corev1.PersistentVolumeClaim, len(cluster.Spec.VolumeClaimTemplates))
-		for i, v := range cluster.Spec.VolumeClaimTemplates {
-			pvc := v.ToCoreV1()
-			pvc.Namespace = cluster.Namespace
-			if err := ctrl.SetControllerReference(cluster, &pvc, r.Scheme); err != nil {
-				panic(err)
-			}
-			pvc.Namespace = ""
-			sts.Spec.VolumeClaimTemplates[i] = pvc
+	mysqldContainer, err := r.makeV1MySQLDContainer(cluster)
+	if err != nil {
+		return err
+	}
+	containers = append(containers, mysqldContainer)
+	containers = append(containers, r.makeV1AgentContainer(cluster))
+
+	if !cluster.Spec.DisableSlowQueryLogContainer {
+		force := cluster.Status.ReconcileInfo.Generation != cluster.Generation
+		sts, err := appsv1ac.ExtractStatefulSet(&orig, fieldManager)
+		if err != nil {
+			return fmt.Errorf("failed to extract StatefulSet: %w", err)
 		}
 
-		sts.Spec.Template.Annotations = mergeMap(sts.Spec.Template.Annotations, cluster.Spec.PodTemplate.Annotations)
-		sts.Spec.Template.Labels = mergeMap(sts.Spec.Template.Labels, cluster.Spec.PodTemplate.Labels)
-		sts.Spec.Template.Labels = mergeMap(sts.Spec.Template.Labels, labelSet(cluster, false))
+		containers = append(containers, r.makeV1SlowQueryLogContainer(sts, force))
+	}
+	if len(cluster.Spec.Collectors) > 0 {
+		containers = append(containers, r.makeV1ExporterContainer(cluster.Spec.Collectors))
+	}
+	containers = append(containers, r.makeV1OptionalContainers(cluster)...)
 
-		podSpec := cluster.Spec.PodTemplate.Spec.DeepCopy()
-		podSpec.ServiceAccountName = cluster.PrefixedName()
+	initContainers := r.makeV1InitContainer(cluster, *mysqldContainer.Image)
```

Comment:
Please add nil check or validation for the `image` field of `mysqld` container. 

```
apiVersion: moco.cybozu.com/v1beta2
kind: MySQLCluster
metadata:
  name: foo
spec:
  podTemplate:
    spec:
      containers:
      - name: mysqld
  volumeClaimTemplates:
  - metadata:
      name: mysql-data
    spec:
      accessModes: [ "ReadWriteOnce" ]
      resources:
        requests:
          storage: 1Gi
```

Reply:
Fixed: https://github.com/cybozu-go/moco/pull/364/commits/ffce0d95e591711bc73051c70f11b8b6e2904561